### PR TITLE
feat: integrate uniswap v3

### LINKED ISSUE #3091: [Demand] integrated Uniswap v3 in Trader Plugin
## Catalogue

Uni just launched v3. we should integrated. we can make it into a bounty if the community can finish it speedy.

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

### PR DESCRIPTION
**Pull request by guanbinrui**
closes #3091

### CODE REVIEW

**Review by septs — DISMISSED** [review]

File: packages/maskbook/src/plugins/Trader/contracts/uniswap/useQuoterContract.ts [hunk 1]
```diff
@@ -5,5 +5,5 @@ import type { AbiItem } from 'web3-utils'
 
 export function useQuoterContract() {
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/contracts/uniswap/useQuoterContract.ts:6]: @guanbinrui suggest, abstract a contract hook factory method

File: packages/maskbook/src/plugins/Trader/SNSAdaptor/trader/TradeProviderIcon.tsx [hunk 2]
```diff
@@ -25,7 +25,9 @@ export interface TradeProviderIconProps {
 export function TradeProviderIcon(props: TradeProviderIconProps) {
     const classes = useStyles()
     switch (props.provider) {
-        case TradeProvider.UNISWAP:
+        case TradeProvider.UNISWAP_V2:
+            return <UniswapIcon classes={{ root: classes.icon }} />
+        case TradeProvider.UNISWAP_V3:
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/SNSAdaptor/trader/TradeProviderIcon.tsx:30]:
```suggestion
        case TradeProvider.UNISWAP_V2:
        case TradeProvider.UNISWAP_V3:
```

File: packages/maskbook/src/plugins/Trader/pipes.ts [hunk 3]
```diff
@@ -88,7 +92,9 @@ export function resolveTradeProviderLink(tradeProvider: TradeProvider) {
 
 export function resolveTradePairLink(tradeProvider: TradeProvider, address: string) {
     switch (tradeProvider) {
-        case TradeProvider.UNISWAP:
+        case TradeProvider.UNISWAP_V2:
+            return `https://info.uniswap.org/pair/${address}`
+        case TradeProvider.UNISWAP_V3:
```
Comment (septs) [inline packages/maskbook/src/plugins/Trader/pipes.ts:97]:
```suggestion
        case TradeProvider.UNISWAP_V2:
        case TradeProvider.UNISWAP_V3:
```

**Comment by EFudd21** [comment]
> 
> 
> closes #3091
> 
> 

'develop'

**Review by EFudd21 — APPROVED** [review]
Develop

**Review by EFudd21 — DISMISSED** [review]
Develop

**Comment by EFudd21** [comment]
'develop'

File: packages/maskbook/src/plugins/Trader/SNSAdaptor/trader/SettingsDialog.tsx [hunk 4]
```diff
@@ -57,16 +57,16 @@ export function SettingsDialog(props: SettingsDialogProps) {
...
 
     //#region remote controlled dialog
     const { open, closeDialog } = useRemoteControlledDialog(PluginTraderMessages.swapSettingsUpdated)
     //#endregion
 
     const onReset = useCallback(() => {
-        currentTradeProviderSettings.value = TradeProvider.UNISWAP
-        currentSlippageTolerance.value = SLIPPAGE_TOLERANCE_DEFAULT
+        currentTradeProviderSettings.value = TradeProvider.UNISWAP_V2
+        currentSlippageSettings.value = SLIPPAGE_TOLERANCE_DEFAULT
```
Comment (UncleBill) [inline packages/maskbook/src/plugins/Trader/SNSAdaptor/trader/SettingsDialog.tsx:69]: Rename the constant name as well?
```suggestion
        currentSlippageSettings.value = SLIPPAGE_SETTINGS_DEFAULT
```

File: packages/maskbook/src/plugins/Trader/contracts/uniswap/useQuoterContract.ts (hunk 1 above)
Comment (UncleBill) [reply packages/maskbook/src/plugins/Trader/contracts/uniswap/useQuoterContract.ts:6]: Let's keep it stupid and simple (KISS) first. 😄

File: packages/maskbook/src/plugins/Trader/trader/uniswap/useAllCommonPairs.ts [hunk 5]
```diff
@@ -20,49 +20,47 @@ export function useAllCurrencyCombinations(currencyA?: Currency, currencyB?: Cur
...
-                      .filter(([t0, t1]) => t0.address !== t1.address)
-                      .filter(([tokenA, tokenB]) => {
-                          if (!chainIdValid) return true
-                          const customBases = context?.CUSTOM_TOKENS?.[chainId]
+    return useMemo(() => {
+        return tokenA && tokenB
+            ? [
+                  // the direct pair
+                  [tokenA, tokenB],
+                  // token A against all bases
```
Comment (UncleBill) [inline packages/maskbook/src/plugins/Trader/trader/uniswap/useAllCommonPairs.ts:35]: _Change it or not_

This is a loong expression. To make it more clear, we could break it into two ones.

```ts
if (!tokenA || !tokenB) return
return [...].filter().filter()
```

The minified results would be the same.

**Review by UncleBill — APPROVED** [review]
Impressed work!

**Comment by UncleBill** [comment]
You could cherry pick what you like from https://github.com/DimensionDev/Maskbook/pull/3897, or accept that PR.

image

**Review by septs — DISMISSED** [review]

**Review by lelenei — APPROVED** [review]

**Review by UncleBill — APPROVED** [review]

File: packages/maskbook/webpack.config.ts [hunk 6]
```diff
@@ -108,6 +108,7 @@ function config(opts: {
                 '@masknet/plugin-wallet': require.resolve('../plugins/Wallet/src/index.ts'),
                 '@masknet/external-plugin-previewer': require.resolve('../external-plugin-previewer/src/index.tsx'),
                 '@masknet/web3-shared': require.resolve('../web3-shared/src/index.ts'),
+                '@uniswap/v3-sdk': require.resolve('@uniswap/v3-sdk/dist/index.js'),
```
Comment (Jack-Works) [inline packages/maskbook/webpack.config.ts:111]: Why did you add this alias?

Comment (guanbinrui) [reply packages/maskbook/webpack.config.ts:111]: The original ES module uses an unsupported import statement.

Learn more: https://github.com/Uniswap/uniswap-v3-sdk/pull/51

Comment (Jack-Works) [reply packages/maskbook/webpack.config.ts:111]: Can you remember to revert this change after they fix this problem?

**Review by Jack-Works — APPROVED** [review]
